### DISCUSSION TITLE
Confusing warning or wrong use of system JDK in the compiler plugin?

### DISCUSSION THREAD

**Discussion by LorenzoBettini** [eclipse-tycho/tycho#3732] [Q&A]
Maybe I don't get how the use of JDK by the compiler plugin should work (I'm using Tycho 4.0.6), but, https://tycho.eclipseprojects.io/doc/latest/tycho-compiler-plugin/compile-mojo.html#useJDK says

> Which JDK to use for compilation. Default value is SYSTEM which means the currently running JDK.

In GitHub Actions I setup Java 21 as the JDK, which is also reported by Maven correctly:

```
Apache Maven 3.9.6 (bc0240f3c744dd6b6ec2920b3cd08dcc295161ae)
Maven home: /home/runner/.m2/wrapper/dists/apache-maven-3.9.6-bin/3311e1d4/apache-maven-3.9.6
Java version: 21.0.2, vendor: Eclipse Adoptium, runtime: /usr/lib/jvm/temurin-21-jdk-amd64
```

So I'd assume Tycho will use this one, since I haven't configured `useJDK`.

However, during the build I see:

```
Warning:  Using JavaSE-17 to fulfill requested profile of JavaSE-11. This might lead to faulty dependency resolution, consider defining a suitable JDK in the toolchains.xml.
```

This comes from this code

https://github.com/eclipse-tycho/tycho/blob/be568d87103984fff0786354e9a78300943c9cc2/tycho-core/src/main/java/org/eclipse/tycho/core/ee/ExecutionEnvironmentUtils.java#L89

which, I seem to understand, searches for JDKs and stops as soon as it finds one higher than the profile of the currently compiled bundle. Since also Java 17 is installed in GitHub Actions (though it's not the set as the default, see above), Tycho is going to use Java 17 instead of Java 21?

By the documentation above I'd assume it should use the JDK set in the system (Java 21).

Or is just a misguided warning message?

Due to some compilation failures later in the build, it looks like it's really using Java 17 instead of Java 21 (the compilation fails because it cannot find the ArrayList.getLast method, introduced in Java 21).

Should I set a toolchain explicitly?

By the way, in our project, Xtext, we still use Java 11 as the required Java, but we'd like to test with Java 21.

thanks in advance

**Comment by laeubi** [comment]
@LorenzoBettini in short, don't use `useJDK` that's probably not useful these days anymore and you most likely don't need toolchains either.

The longer story is `useJDK` _enforces_ (e.g. fails otherwise) to use a JDK matching the BREE requirement of the bundle, so if it would be `JavaSE-8` it requires to have a "real" Java 8 JVM used for compilation and it requires to define such in `toolchains.xml` because otherwise it won't work to use "real" Java 8 with "real" Java 11 and so on in the same build.

For you use-case, simply make either only Java 21 available and Tycho will pass the appropriate target level to the compiler and everything works fine. And even the warning should be gone with Tycho 4.0.7 now because its not really useful.

If compilation fails, it is most likely because your bundle is still at BREE 11 where it would require BREE 21 to use Java 21 APIs.

**Comment by LorenzoBettini** [comment]
@laeubi I confirm that the compilation failure was due to another issue.

However, does the warning

```
Warning:  Using JavaSE-17 to fulfill requested profile of JavaSE-11. This might lead to faulty dependency resolution, consider defining a suitable JDK in the toolchains.xml.
```

means that JDK 17 will be effectively used instead of JDK 21?

As I said, in GitHub Action you can set a JDK to be the default, but, in any case, JDK 8, 11 and 17 are still available in the system (and cannot be uninstalled as far as I know). So, does the warning above say that JDK 17 will be used since it can be found in the system?

For the moment, I make sure to use JDK 21 with the maven toolchain plugin.

**Reply by laeubi** [reply]
> However, does the warning
> ...
> means that JDK 17 will be effectively used instead of JDK 21?

The warning means, that somewhere in your build you have a project that uses BREE 11, but not configured any toolchain JDK for Java 11 therefore it will use the "next higher" (here one configured for Java 17) to compute the available system-packages.

It will still use Java 21 to compile the code but using `--target 11` as the compiler option.

**Comment by LorenzoBettini** [comment]
And in the Tycho 4 branch the log and check seem to be still there:

https://github.com/eclipse-tycho/tycho/blob/tycho-4.0.x/tycho-core/src/main/java/org/eclipse/tycho/core/ee/ExecutionEnvironmentUtils.java#L93

**Reply by laeubi** [reply]
> And in the Tycho 4 branch the log and check seem to be still there:

If you look close enough it is now only a **debug** message :)

**Comment by travkin79** [comment]
Hello @laeubi,

> in short, don't use useJDK that's probably not useful these days anymore and you most likely don't need toolchains either.

Since I get the following warning message when using tycho 4.0.8 in combination with `<useJdk>BREE</useJdk>`, I guess, the parameter _useJdk_ became obsolete. Is that right? If that's the case, the [tycho documentation](https://tycho.eclipseprojects.io/doc/latest/tycho-compiler-plugin/compile-mojo.html#useJDK) should be updated.

```
[WARNING] Parameter 'useJdk' is unknown for plugin 'tycho-compiler-plugin:4.0.8:validate-classpath ( default-validate-classpath ) '
[WARNING] Parameter 'useJdk' is unknown for plugin 'tycho-compiler-plugin:4.0.8:compile ( default-compile ) '
[WARNING] Parameter 'useJdk' is unknown for plugin 'tycho-compiler-plugin:4.0.8:testCompile ( default-testCompile ) '
```

**Reply by laeubi** [reply]
As you can see the parameter is still there (otherwise the doc wont mention it) so it looks like a configuration issue or a bug in maven.
